### PR TITLE
Rust: extract crate graph

### PR DESCRIPTION
**Pull request by aibaars**
This pull request changes the Rust extractor to emit information about the crate graph, and the  public modules, types, traits and function signatures contained in each crate. This should give our analyses information about names and types provided by the dependencies of a project.  The information in the crate graph is represented as AST nodes without macro calls; type paths should be fully qualified and methods/functions have no bodies.

This PR currently does not extract generic type arguments , `where` clauses, and associated types and aliases. These should be added in a follow-up pull request.

### CODE REVIEW

File: rust/extractor/src/crate_graph.rs [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+    for krate_id in crate_graph.iter() {
+        if let Some((root_module_file, hash)) = crate_id_map.get(&krate_id) {
+            let path = root_module_file.join(format!("{hash:0>16x}"));
+            let mut trap = trap_provider.create("crates", path.as_path());
+            if trap.path.exists() {
+                continue;
+            }
+            let krate = &crate_graph[krate_id];
+            let element = generated::Crate {
+                id: trap::TrapId::Key(format!("crate:{}:{hash}", root_module_file.display())),
```
Comment (redsun82) [inline rust/extractor/src/crate_graph.rs:71]: the `crate:` prefix is not really required, as we already prefix trap keys with the type name that we use it for (in this case `Crate_`)

File: rust/extractor/src/crate_graph.rs [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+            crate_id_map.insert(krate_id, (root_module_file, hash));
+        }
+    }
+    for krate_id in crate_graph.iter() {
+        if let Some((root_module_file, hash)) = crate_id_map.get(&krate_id) {
+            let path = root_module_file.join(format!("{hash:0>16x}"));
+            let mut trap = trap_provider.create("crates", path.as_path());
+            if trap.path.exists() {
+                continue;
+            }
```
Comment (redsun82) [inline rust/extractor/src/crate_graph.rs:68]: maybe we need a comment here that this is a potential race condition if we ever start doing this in parallel...

File: rust/extractor/src/crate_graph.rs [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+            );
+            trap.commit().unwrap_or_else(|err| {
+                log::error!(
+                    "Failed to write trap file for crate: {}: {}",
+                    root_module_file.display(),
+                    err.to_string()
+                )
+            });
+        }
+        fn go(
```
Comment (redsun82) [inline rust/extractor/src/crate_graph.rs:111]: I think we need a little more structure here, and extract parts of this function to smaller pieces. Right now it's very hard to follow what is happening here, this whole `extract_crate_graph` function body is just too big

File: rust/extractor/src/crate_graph.rs [hunk 4]
```diff
@@ -0,0 +1,1324 @@
...
+                    start_column: 0,
+                    end_line: 0,
+                    end_column: 0,
+                },
+            );
+            trap.writer.add_tuple(
+                "locatable_locations",
+                vec![module.into(), location_label.into()],
+            );
+
```
Comment (redsun82) [inline rust/extractor/src/crate_graph.rs:104]: Nit: I'd say this block may be moved to the `emit_module` function, as I'd say it's part of the "module" emission

**Review by redsun82 — APPROVED** [review]
That's quite the tremendous work, thanks!!

I only have a minor nit, feel free to ignore.

File: rust/ql/test/extractor-tests/crate_graph/modules.expected [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+#-----| fn length
+
+#-----| impl ...::Display for ...::X { ... }
+#-----|  -> fn fmt
+
+#-----| impl AsString for ...::X { ... }
+#-----|  -> fn as_string
+
+lib.rs:
+#    1| mod crate
```
Comment (aibaars) [inline rust/ql/test/extractor-tests/crate_graph/modules.expected:22]:
```suggestion
#    0| mod crate
```

File: rust/extractor/src/crate_graph.rs (hunk 4 above)
Comment (redsun82) [reply rust/extractor/src/crate_graph.rs:104]: ah, I had missed that! Maybe then we could rename `module` to `root_module` for extra clarity? Also, it may make sense to add a `fn emit_file_only_location(&mut self, file_label: Label<generated::File>, entity_label: Label<E>)` to `TrapFile` doing this whole block (and maybe internally reusing some of the code in `emit_location` and `emit_location_label`, so that here we could just write
```suggestion
            let file_label = trap.emit_file(root_module_file);
            trap.emit_file_only_location(root_module, file_label);
```

Comment (redsun82) [reply rust/extractor/src/crate_graph.rs:104]: but that's ok also left for follow-up changes.

**Review by redsun82 — APPROVED** [review]

File: rust/ql/lib/codeql/files/FileSystem.qll [hunk 6]
```diff
@@ -55,7 +55,8 @@ class File extends Container, Impl::File {
         |
           node.getFile() = this and
           line = [/*loc.getStartLine(), */ loc.getEndLine()] and // ignore start locations for now as we're getting them wrong for things with a comment attached
-          not loc instanceof EmptyLocation
+          not loc instanceof EmptyLocation and
+          line > 0
```
[the remark this answers is not in the document]
Comment (aibaars) [reply rust/ql/lib/codeql/files/FileSystem.qll:59]: Lines start at `1`. A location with a `0` line is pointing to a whole file .

File: rust/ql/lib/codeql/rust/internal/AstConsistency.qll [hunk 7]
```diff
@@ -24,7 +24,9 @@ query predicate multipleLocations(Locatable e) { strictcount(e.getLocation()) >
 /**
  * Holds if `e` does not have a `Location`.
  */
-query predicate noLocation(Locatable e) { not exists(e.getLocation()) }
+query predicate noLocation(Locatable e) {
+  not exists(e.getLocation()) and not e.(AstNode).getParentNode*() = any(Crate c).getModule()
```
[the remark this answers is not in the document]
Comment (aibaars) [reply rust/ql/lib/codeql/rust/internal/AstConsistency.qll:28]: No, we have crate information for things in source code, but the crate information has no locations.

[the remark this answers is not in the document]
Comment (hvitved) [reply rust/ql/lib/codeql/rust/internal/AstConsistency.qll:28]: Just to double check: This is because for things in source code, we will now extract two entities, one as before, and a new one with crate information (for which we do not want to report missing locations)?

[the remark this answers is not in the document]
Comment (aibaars) [reply rust/ql/lib/codeql/rust/internal/AstConsistency.qll:28]: Yes, that's right. For items in source code, you'd have signatures without bodies and without locations.

**Review by hvitved — APPROVED** [review]
